src/App.tsx

```diff
@@ -1,24 +1,195 @@
+        currentTodos => [...currentTodos, response],
+      ))
+      .then(() => setTempTodo(null))
+      .catch(() => {
+        setErrorMessage(Errors.UnableToAddATodo);
+        throw new Error(Errors.UnableToAddATodo);
+      })
+      .finally(
+        () => setLoadingTodosPause(prevIds => prevIds
+          .filter(prevId => prevId !== 0)),
+      );
+  }, [errorMessage]);
+
+  function deleteTodo(todoId: number) {
+    setLoadingTodosPause(prev => [...prev, todoId]);
+
+    return todosApiServices.deleteUserTodo(todoId)
+      .then(() => setTodos(currentTodos => currentTodos
+        .filter(todo => todo.id !== todoId)))
+      .catch((error) => {
+        setTodos(todos);
+        setErrorMessage(Errors.UnabletoDeleteATodo);
+        throw error;
+      })
+      .finally(() => {
+        setLoadingTodosPause(
+          prev => prev.filter(loadingId => loadingId !== todoId),
+        );
+      });
+  }
+
+  async function clearCompletedTodos() {
+    const completedTodos
+      = todos?.filter(({ completed }) => completed);
+
+    if (completedTodos?.length) {
+      Promise.all(
+        completedTodos.map((todo) => deleteTodo(todo.id)),
+      )
+        .catch(() => {
+          setErrorMessage(Errors.UnabletoDeleteATodo);
+        });
+    }
+  }
+
+  function updateTodo(updatedTodo: Todo) {
+    setLoadingTodosPause(currentTodo => [...currentTodo, updatedTodo.id]);
+
+    return todosApiServices.updateTodo(updatedTodo)
+      .then(() => setTodos(prev => (
+        prev.map(prevTodo => (
+          prevTodo.id === updatedTodo.id
+            ? updatedTodo
+            : prevTodo
+        ))
+      )))
+      .catch(() => setErrorMessage(Errors.UnableToUpdateATodo))
+      .finally(() => {
+        setLoadingTodosPause(currentTodos => currentTodos
+          .filter(todoId => updatedTodo.id !== todoId));
+      });
+  }
+
+  async function toggleAll() {
+    const isAllCompleted = todos
+      .every(todo => todo.completed);
+
+    const updatedTodos = todos
+      .filter(todo => isAllCompleted === todo.completed);
+
+    await Promise.all(updatedTodos.map(todo => (
+      updateTodo({
+        ...todo,
+        completed: !isAllCompleted,
+      })
+    )));
   }
 
   return (
-    <section className="section container">
-      <p className="title is-4">
-        Copy all you need from the prev task:
-        <br />
-        <a href="https://github.com/mate-academy/react_todo-app-add-and-delete#react-todo-app-add-and-delete">React Todo App - Add and Delete</a>
-      </p>
-
-      <p className="subtitle">Styles are already copied</p>
-    </section>
+    <div className="todoapp">
+      <h1 className="todoapp__title">todos</h1>
+
+      <div className="todoapp__content">
+        <header className="todoapp__header">
+          {/* this buttons is active only if there are some active todos */}
+          {!!todos.length && (
+            <ToggleAll
+              onToggleAll={() => toggleAll()}
```

Comment:
No need in additional callback cuz of lint) - fix other places
```suggestion
              onToggleAll={toggleAll}
```